### DISCUSSION TITLE
next.js plugin is missing types

### DISCUSSION THREAD

**Discussion by stefanprobst** [markdoc/markdoc#454] [Bug report] [answered]
### What happened?

when importing the next.js plugin with `import createMarkdocPlugin from "@markdoc/next.js";` and trying to call it, i get a "This expression is not callable." type error.

### To reproduce

add this to `next.config.mjs`:

```ts
//@ts-check

import createMarkdocPlugin from "@markdoc/next.js";

const withMarkdoc = createMarkdocPlugin()

/** @type {import('next').NextConfig} */
const nextConfig = {}

// EDIT:
export default withMarkdoc(nextConfig)
```

### Additional context

needs something like this:

```ts
type WithMarkdoc = (config: NextConfig) => NextConfig

declare function createMarkdocPlugin(options?: Options): WithMarkdoc

export = createMarkdocPlugin
```

**Comment by mfix-stripe** [comment]
Thanks for sharing @stefanprobst. 

Can you try:

```js
const withMarkdoc = require('@markdoc/next.js');
```
instead?

**Reply by stefanprobst** [reply]
@mfix-stripe sorry for the typo in the example. but both

`next.config.cjs` with:

```ts
//@ts-check

const createMarkdocPlugin = require('@markdoc/next.js');

const withMarkdoc = createMarkdocPlugin()

/** @type {import('next').NextConfig} */
const nextConfig = {}

module.exports = withMarkdoc(nextConfig)
```

and `next.config.mjs` with

```ts
//@ts-check

import createMarkdocPlugin from "@markdoc/next.js";

const withMarkdoc = createMarkdocPlugin()

/** @type {import('next').NextConfig} */
const nextConfig = {}

export default withMarkdoc(nextConfig)
```

show this error.

why convert this to a discussion?

**Reply by mfix-stripe** [reply, marked as answer]
What you want is:

```js
const withMarkdoc = require('@markdoc/next.js');

module.exports = withMarkdoc()({
  ...
});
```

See https://github.com/markdoc/docs/blob/main/next.config.js for an example.

**Reply by mfix-stripe** [reply]
`@markdoc/next.js` isn't written in typescript, so I'm not sure what types you are actually getting.

**Reply by stefanprobst** [reply]
> I'm not sure what types you are actually getting.

@mfix-stripe this is what this issue is about! for this plugin to not give a ts error with `@ts-check` or `checkJs: true`, you should provide a `index.d.ts` file in this repository. it should contain something like:

```ts
type WithMarkdoc = (config: NextConfig) => NextConfig

declare function createMarkdocPlugin(options?: Options): WithMarkdoc

export = createMarkdocPlugin
```

this is also how other next.js plugins do it. see for example the official mdx plugin here: https://github.com/vercel/next.js/blob/canary/packages/next-mdx/index.d.ts

**Reply by mfix-stripe** [reply]
Gotcha gotcha. @stefanprobst  Can you take a look at this PR and see if that works for your use case when you get a chance? https://github.com/markdoc/next.js/pull/44

**Reply by stefanprobst** [reply]
cool, thanks! i'll take a look!
